File: packages/axon-core-ts/src/pack.ts
/** Battery hierarchy: Cell → Module → Pack. */

export interface Cell {
  id: string;
  serial?: string;
  lotId: string;
  chemistry: string; // e.g., "NMC811", "LFP", "NCA"
  formatCode: string; // e.g., "21700", "4680", "PHEV2"
  capacityNominal_Ah: number;
  capacityMeasured_Ah?: number;
  ocv_V?: number;
  internalResistance_mOhm?: number;
  productionOrderId?: string;
}

export interface Module {
  id: string;
  serial: string;
  cellIds: string[];
  topology: string; // e.g., "12s2p"
  productionOrderId?: string;
}

export interface Pack {
  id: string;
  serial: string;
  moduleIds: string[];
  customer?: string;
  application: "EV" | "ESS" | "industrial" | "marine" | "aerospace" | "other";
  bmsHardwareRev?: string;
  bmsFirmwareRev?: string;
  productionOrderId?: string;
}

File: packages/axon-core-py/src/axon_core/pack.py
from __future__ import annotations
from typing import Literal, Optional
from pydantic import BaseModel, Field


class Cell(BaseModel):
    id: str
    serial: Optional[str] = None
    lotId: str
    chemistry: str
    formatCode: str
    capacityNominal_Ah: float
    capacityMeasured_Ah: Optional[float] = None
    ocv_V: Optional[float] = None
    internalResistance_mOhm: Optional[float] = None
    productionOrderId: Optional[str] = None


class Module(BaseModel):
    id: str
    serial: str
    cellIds: list[str] = Field(default_factory=list)
    topology: str
    productionOrderId: Optional[str] = None


class Pack(BaseModel):
    id: str
    serial: str
    moduleIds: list[str] = Field(default_factory=list)
    customer: Optional[str] = None
    application: Literal["EV", "ESS", "industrial", "marine", "aerospace", "other"]
    bmsHardwareRev: Optional[str] = None
    bmsFirmwareRev: Optional[str] = None
    productionOrderId: Optional[str] = None
